Upgrade to Celery 4

EnTeQuAk opened the pull request:
Fixes #5549

Finally brings our code-base on par with celery 4 which brings TONS of cleanups, chord/group fixes and as far as I can tell should be lots more stable / easier now.

See docs.celeryproject.org/en/latest/whatsnew-4.0.html for a few more details.

There are not much differences in how chords/groups work in many details. Nested groups get flattened but we don't use that and it's useful (semantically at least, that's an optimization - it's how they worked anyway).

r? @eviljeff or @diox 

This is in preparation to be up-to-date before working more on django 1.11 and give give process-request-task another go later and profit of all those new cleanups.

cc @bqbn, there are not much changes in how celery is being used from a ops perspective as far as I can tell but it'd be nice to get familiar with the "what changed" docs and generally that this is happening in case we run into any problems.

What changed is the environment variable for the celery broker url will change from `BROKER_URL` to `CELERY_BROKER_URL`.

EnTeQuAk commented on .travis.yml:
```diff
@@ -28,6 +28,7 @@ services:
   - mysql
   - memcached
   - elasticsearch
+  - redis
```
Celery now runs eager tasks through their broker + result backend accordingly so we need redis on travis now.

This doesn't mean that "always eager" is encouraged now, it's still discouraged but we are using it so heavily that I didn't want to change that just now.

eviljeff commented on src/olympia/addons/cron.py:
```diff
@@ -8,7 +8,7 @@
 
 import multidb
-from celery.task.sets import TaskSet
+from celery import group
```
is `group` a drop-in replacement for `TaskSet`?

EnTeQuAk replied:
Correct, it is.

eviljeff commented on src/olympia/amo/tests/test_send_mail.py:
```diff
@@ -280,10 +281,12 @@ def test_async_will_retry(self, backend):
+            send_mail('test subject',
+                      'test body',
+                      async=True,
+                      recipient_list=['somebody@mozilla.org'])
```
Not sure I understand this replacement - so now if there's a `RuntimeError` on the first try it fails on the second try?  It never succeeds?

EnTeQuAk replied:
It does but now it raises a `Retry` exception that get's handled by the worker and retried properly. This didn't happen before for various things that masked that behavior and actually retried running the task during the tests. But we were never able to see that, now we are.

EnTeQuAk replied:
afaik this changed with https://github.com/celery/celery/pull/2607/commits/7bf736bfae86766819c3af5da661f7152b364c85

eviljeff commented on src/olympia/bandwagon/cron.py:
```diff
@@ -3,7 +3,8 @@
 from django.db.models import Count
 
-from celery.task.sets import TaskSet
+
```
extra line

eviljeff commented on src/olympia/conf/prod/settings.py:
```diff
@@ -85,11 +85,11 @@
-BROKER_CONNECTION_TIMEOUT = 0.5
+CELERY_TASK_IGNORE_RESULT = True
+CELERY_WORKER_DISABLE_RATE_LIMITS = True
+CELERY_BROKER_CONNECTION_TIMEOUT = 0.5
```
nit: I think some of the above settings are redundant because they're the same as `settings_base`.  Eventually it would be nice if the `/conf/` files _only_ contained settings that differ from `settings_base`

EnTeQuAk replied:
indeed, there's an issue somewhere but I can't find it right now. I'd rather fix this in a separate pull request and unify things across all boards.

EnTeQuAk commented:
@bqbn What changed is the environment variable for the celery broker url will change from BROKER_URL to CELERY_BROKER_URL. Can we get this prepared so if this PR get's merged I can ping you to propagate those changes? For the time being you can easily set both variables too so that you don't have to wait for me.

EnTeQuAk commented:
Just a few minor changes since the last review - for the sake of getting it - re-r? @eviljeff?

@bqbn will prepare the config change today (wed) to propagate with this merge on dev, stage and next week prod so once this is done we can merge it.

eviljeff commented on docker/supervisor-celery.conf:
```diff
@@ -8,7 +8,7 @@ logfile=/code/logs/supervisord-celery.log
 # - Set concurrency (number of workers) back to the default
 # - The code auto-reloads for development
-command=celery -A olympia.amo.celery:app worker --autoreload -E --loglevel=INFO
+command=watchmedo auto-restart -d src/ -p '*.py' -- celery -A olympia.amo.celery:app worker -E --loglevel=INFO
```
what's this for?

EnTeQuAk replied:
Celery 4 removed it's autoreload feature because it was buggy and unmaintained. watchdog replaces that for us now.

eviljeff commented on docker-compose.yml:
```diff
@@ -72,3 +72,5 @@ services:
     links:
      - "web:olympia.dev"
+    extra_hosts:
+     - "olympia.dev:web"
```
this is a weird, weird, pattern.  `links` is really the opposite of `extra_hosts`?

eviljeff replied:
oh, you removed it in a later commit, nevermind.

bqbn commented:
If this is ready, I recommend we merge it after tomorrow's (Thursday) push.

EnTeQuAk commented:
Cool, sounds like a plan. Let's do that 


On Wed, Oct 11, 2017, at 09:54 PM, bqbn wrote:
> If this is ready, I recommend we merge it after tomorrow's
> (Thursday) push.> — You are receiving this because you authored the thread. Reply to
> this email directly, view it on GitHub[1], or mute the thread[2].> 

Links:

  1. https://github.com/mozilla/addons-server/pull/6625#issuecomment-335928458
  2. https://github.com/notifications/unsubscribe-auth/AAIfGa6Jf0Ag05qNy3YiapSKFeTj-uN1ks5srRzYgaJpZM4Pyvwb
